bpf/kprobe_pwru.c

```diff
@@ -341,10 +341,31 @@ PWRU_ADD_KPROBE(5)
 SEC("kprobe/skb_lifetime_termination")
 int kprobe_skb_lifetime_termination(struct pt_regs *ctx) {
 	u64 skb = (u64) PT_REGS_PARM1(ctx);
-	if (cfg->track_skb)
-		bpf_map_delete_elem(&skb_addresses, &skb);
 
-	return 0;
+	bpf_map_delete_elem(&skb_addresses, &skb);
+
+	return BPF_OK;
+}
+
+static __always_inline int
+track_skb_clone(struct pt_regs *ctx) {
+	u64 new = (u64) PT_REGS_PARM1(ctx);
+	u64 old = (u64) PT_REGS_PARM2(ctx);
```

Comment:
But, we're unable to get the `skb` argument in `kretprobe`.

As your mention, `fexit` is better. But 5.4 kernel does not support it.

I like `fexit` too. Because it can track skb clones accurately.

Reply:
Using something unchanged between kp and krp, we can store old skb at kp then look it up at krp, something like:
```
k:skb_clone {@map[$unchanged] = $arg1; }
kr:skb_clone {$old = (struct sk_buff*) @map[$unchanged]; }
```

We have multi options for $unchanged var. For me, I used `$tid` a lot although I'm not sure it's guaranteed not to change.
If you're not sure about tid, use `$rsp@k == $rsp-8@kr` to link two events.

Reply:
Hi @brb , what do you think? 

Should we add a --follow-skb-clones option and then implement with `fexit`?

Or implement with kp and krp, suggested by Gray?

Or the way of this PR?

Reply:
Hey, considering how >= 5.4 is widespread and that pwru requires >= 5.3, I'd suggest to implement it as `fexit`, and enable it if `--filter-track-skb` is set and a kernel supports `fexit`. I'd rather keep a set of possible args small (better UX).